smp: Move for loops to use arch_num_cpus instead of CONFIG_MP_NUM_CPUS

galak opened the pull request:
Change for loops of the form:

for (i = 0; i < CONFIG_MP_NUM_CPUS; i++)
   ...

to

unsigned int num_cpus = arch_num_cpus();
for (i = 0; i < num_cpus; i++)
   ...

We do the call outside of the for loop so that it only happens once, rather than on every iteration.

Signed-off-by: Kumar Gala <kumar.gala@intel.com>

galak committed:
smp: Move for loops to use arch_num_cpus instead of CONFIG_MP_NUM_CPUS

Change for loops of the form:

for (i = 0; i < CONFIG_MP_NUM_CPUS; i++)
   ...

to

unsigned int num_cpus = arch_num_cpus();
for (i = 0; i < num_cpus; i++)
   ...

We do the call outside of the for loop so that it only happens once,
rather than on every iteration.

Signed-off-by: Kumar Gala <kumar.gala@intel.com>

teburd reviewed (changes requested):
Looks good, the signed off line in the first commit probably should match the second one though.

galak commented:
> Looks good, the signed off line in the first commit probably should match the second one though.

Oh, good catch.  ;).

galak committed:
smp: define arch_num_cpus always

Move arch_num_cpus outside of CONFIG_SMP in sys/arch_interface.h and
add define arch_num_cpus on all platforms (added arch_inlines.h on
those platforms that didn't have it before).

Signed-off-by: Kumar Gala <kumar.gala@intel.com>

carlocaione reviewed (approved):
LGTM. I wonder if an helper like `FOR_EACH_CPU()` could be more handy here instead of boilerplating.

thoh-ot commented on include/zephyr/arch/arch_inlines.h:
```diff
@@ -24,6 +24,16 @@
+#elif defined(CONFIG_SPARC)
+#include <zephyr/arch/sparc/arch_inlines.h>
+#else
+#error "Unknown Architecture"
```
I know i'm a bit late to the party here, but any thoughts on how this is supposed to work with out-of-tree proprietary architectures (such as the one our company is currently using)?

We don't have SMP support so we don't really care about the CONFIG_MP_MAX_NUM_CPUS, but we do care that we now get an "Unknown Architecture" error.

Could we agree to revert the #error change?
